ignite/pkg/protoanalysis/protoutil/cursor.go

```diff
@@ -0,0 +1,302 @@
+package protoutil
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/emicklei/proto"
+)
+
+// Note: The traversing can also be done with proto.Walk but there's some reasons
+// why I chose the cursor instead:
+//
+// 1. We can abort traversing deeper in the tree at any point. (a post
+//    ApplyFunc returning false)
+// 2. We keep track of the parent to have finer grained control over where we
+//    are in the tree.
+// 3. We can use pre/post handling.
+
+// Modeled heavily after the Apply/Cursor logic in astutil, using proto.Visitee as
+// the common interface, abilities for reflection aren't as rich but can still
+// manage in order to get the job done.
+// Cursor has been augmented to add a couple more methods that can make
+// our life easier.
+
+// An ApplyFunc is invoked by Apply for each Visitee n, even if n is nil,
+// before and/or after the node's children, using a Cursor describing
+// the current node and providing operations on it.
+//
+// The return value of ApplyFunc controls the syntax tree traversal.
+// See Apply for details.
+type ApplyFunc func(*Cursor) bool
+
+func Apply(root proto.Visitee, pre, post ApplyFunc) (result proto.Visitee) {
+	parent := &struct{ proto.Visitee }{root}
```

Comment:
Looks like black magic ^^'

So `parent` is an implementation of `proto.Visitee` that embeds `root`, right ? 

I'm not sure to understand how is it different from `root` ? I mean, as far as I understand, calling `parent`'s methods is like calling `root`'s methods, so I'm confused about the usage of `parent`.

Reply:
So yeah, this just initializes the parent. In general, in order to replace a given node while traversing the proto AST, you need a handle to the parent. This statement just initializes the parent with the initial root (whatever that might be). This was, we can replace `root` too if required (I don't think there's an instance of that currently).

Reply:
OK I think I got it, thx.